homeassistant/components/media_player/firetv.py

```diff
@@ -4,166 +4,147 @@
     vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
-    vol.Optional(CONF_SSL, default=DEFAULT_SSL): cv.boolean,
+    vol.Optional(CONF_ADBKEY): has_adb_files,
+    vol.Optional(CONF_GET_SOURCE, default=DEFAULT_GET_SOURCE): cv.boolean,
+    vol.Optional(CONF_GET_SOURCES, default=DEFAULT_GET_SOURCES): cv.boolean
 })
 
+PACKAGE_LAUNCHER = "com.amazon.tv.launcher"
+PACKAGE_SETTINGS = "com.amazon.tv.settings"
+
 
 def setup_platform(hass, config, add_entities, discovery_info=None):
     """Set up the FireTV platform."""
-    name = config.get(CONF_NAME)
-    ssl = config.get(CONF_SSL)
-    proto = 'https' if ssl else 'http'
-    host = config.get(CONF_HOST)
-    port = config.get(CONF_PORT)
-    device_id = config.get(CONF_DEVICE)
-
-    try:
-        response = requests.get(
-            DEVICE_LIST_URL.format(proto, host, port)).json()
-        if device_id in response[CONF_DEVICES].keys():
-            add_entities([FireTVDevice(proto, host, port, device_id, name)])
-            _LOGGER.info("Device %s accessible and ready for control",
-                         device_id)
-        else:
-            _LOGGER.warning("Device %s is not registered with firetv-server",
-                            device_id)
-    except requests.exceptions.RequestException:
-        _LOGGER.error("Could not connect to firetv-server at %s", host)
-
-
-class FireTV:
-    """The firetv-server client.
-
-    Should a native Python 3 ADB module become available, python-firetv can
-    support Python 3, it can be added as a dependency, and this class can be
-    dispensed of.
-
-    For now, it acts as a client to the firetv-server HTTP server (which must
-    be running via Python 2).
-    """
+    from firetv import FireTV
+
+    host = '{0}:{1}'.format(config[CONF_HOST], config[CONF_PORT])
+
+    if CONF_ADBKEY in config:
+        ftv = FireTV(host, config[CONF_ADBKEY])
+        adb_log = " using adbkey='{0}'".format(config[CONF_ADBKEY])
+    else:
+        ftv = FireTV(host)
+        adb_log = ""
+
+    if not ftv.available:
+        _LOGGER.warning("Could not connect to Fire TV at %s%s", host, adb_log)
+        return None
+
+    name = config[CONF_NAME]
+    get_source = config[CONF_GET_SOURCE]
+    get_sources = config[CONF_GET_SOURCES]
+
+    device = FireTVDevice(ftv, name, get_source, get_sources)
+    add_entities([device])
+    _LOGGER.info("Setup Fire TV at %s%s", host, adb_log)
+
+
+def adb_decorator(override_available=False):
+    """Send an ADB command if the device is available and not locked."""
+    def adb_wrapper(func):
+        """Wait if previous ADB commands haven't finished."""
+        @functools.wraps(func)
+        def _adb_wrapper(self, *args, **kwargs):
+            # If the device is unavailable, don't do anything
+            if not self.available and not override_available:
+                return None
+
+            # If an ADB command is already running, skip this command
+            if self.adb_lock:
```

Comment:
```py
if not self.adb_lock.acquire():
    ...
    return None
...
finally:
    self.adb_lock.release()
```

Reply:
Fixed in https://github.com/home-assistant/home-assistant/pull/17767/commits/e14e686b6937c969773fa580bd8b7200664e90d4